testdata/p4_16_samples/hashext.p4

```diff
@@ -0,0 +1,38 @@
+/*
+Copyright 2018 Barefoot Networks, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+extern hash_function<O> {
```

Comment:
Maybe this works better if you make it `hash_function<O, T>`?

Reply:
Unfortunately, that doesn't really make sense -- a hash algorithm should be polymorphic over different amounts/types of inputs -- a single instantiation of this extern should not be tied to a particular input and might in fact be used fo mulitple different inputs (though an architecture might restrict it dynamically to a single use per packet).

Reply:
Actually I think it does make sense: think hardware. You will need to allocate resources for the specific types and alignment of these values that you need to access. So you will need different instances for different hashes.
Why do you think it is a problem to require people to allocate an instance for each call?